sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -556,6 +574,46 @@ private static void verifyUnsplittableMethods(ErrorReporter errors, DoFnSignatur
     return DoFnSignature.SplitRestrictionMethod.create(m, restrictionT);
   }
 
+  private static ImmutableMap<String, TimerDeclaration> analyzeTimerDeclarations(
+      ErrorReporter errors, Class<?> fnClazz) {
+    Map<String, DoFnSignature.TimerDeclaration> declarations = new HashMap<>();
+    for (Field field : declaredFieldsWithAnnotation(DoFn.TimerId.class, fnClazz, DoFn.class)) {
+      String id = field.getAnnotation(DoFn.TimerId.class).value();
+
+      if (declarations.containsKey(id)) {
+        errors.throwIllegalArgument(
+            "Duplicate %s \"%s\", used on both of [%s] and [%s]",
+            DoFn.TimerId.class.getSimpleName(),
+            id,
+            field.toString(),
+            declarations.get(id).field().toString());
+        continue;
+      }
+
+      Class<?> timerSpecRawType = field.getType();
+      if (!(timerSpecRawType.equals(TimerSpec.class))) {
+        errors.throwIllegalArgument(
+            "%s annotation on non-%s field [%s] that has class %s",
+            DoFn.TimerId.class.getSimpleName(),
+            TimerSpec.class.getSimpleName(),
+            field.toString(),
+            timerSpecRawType.getName());
+        continue;
+      }
+
+      if (!Modifier.isFinal(field.getModifiers())) {
+        errors.throwIllegalArgument(
+            "Non-final field %s annotated with %s. Timer declarations must be final.",
+            field.toString(), DoFn.TimerId.class.getSimpleName());
+        continue;
+      }
+
+      declarations.put(id, DoFnSignature.TimerDeclaration.create(id, field));
```

Comment:
I assume you're planning on validating the callbacks (and existence of them) in a future PR. If not, it should be done here.